File: labs/control-ui/api-shadow/upstream_matrix.ts
export interface MatrixUpstreamNode {
  id: string;
  protocol: string;
  endpoint: string;
  latencyMs: number;
  isAlive: boolean;
}

export class UpstreamMatrixUI {
  private nodes = new Map<string, MatrixUpstreamNode>();

  register(node: MatrixUpstreamNode): void {
    this.nodes.set(node.id, { ...node });
  }

  updateHealth(id: string, isAlive: boolean, latencyMs: number): void {
    const node = this.nodes.get(id);
    if (node) {
      node.isAlive = isAlive;
      node.latencyMs = latencyMs;
    }
  }

  selectBest(): MatrixUpstreamNode | null {
    const alive = Array.from(this.nodes.values()).filter((n) => n.isAlive);
    if (alive.length === 0) return null;
    alive.sort((a, b) => a.latencyMs - b.latencyMs);
    return alive[0]!;
  }
}

File: labs/mobile/donor-archive/tunnel/UpstreamMatrix.kt
package com.luminet.android.tunnel

data class AndroidUpstreamNode(
    val id: String,
    val protocol: String,
    val endpoint: String,
    var latencyMs: Long,
    var isAlive: Boolean
)

class UpstreamMatrix {
    private val nodes = mutableMapOf<String, AndroidUpstreamNode>()

    fun register(node: AndroidUpstreamNode) {
        nodes[node.id] = node
    }

    fun update(id: String, isAlive: Boolean, latencyMs: Long) {
        nodes[id]?.let {
            it.isAlive = isAlive
            it.latencyMs = latencyMs
        }
    }

    fun selectBest(): AndroidUpstreamNode? {
        return nodes.values
            .filter { it.isAlive }
            .minByOrNull { it.latencyMs }
    }
}
